pyhap/accessory_driver.py

```diff
@@ -859,118 +860,99 @@ def set_characteristics(self, chars_query, client_addr):
         :type chars_query: dict
         """
         # TODO: Add support for chars that do no support notifications.
-        updates = {}
-        setter_results = {}
-        setter_responses = {}
-        had_error = False
-        had_write_response = False
-        expired = False
 
-        if HAP_REPR_PID in chars_query:
-            pid = chars_query[HAP_REPR_PID]
+        queries = chars_query.get(HAP_REPR_CHARS, [])
+
+        to_notify = (query for query in queries if HAP_PERMISSION_NOTIFY in query)
+        for query in to_notify:
+            aid, iid, ev = query[HAP_REPR_AID], query[HAP_REPR_IID], query[HAP_PERMISSION_NOTIFY]
+            char_topic = get_topic(aid, iid)
+            action = "Subscribed" if ev else "Unsubscribed"
+            logger.debug(
+                "%s client %s to topic %s", action, client_addr, char_topic
+            )
+            self.async_subscribe_client_topic(
+                client_addr, char_topic, ev
+            )
```

Comment:
This looks fully self contained. I think it could be broken out into another function to make this function smaller and easier to read

Reply:
Makes sense. Protected class method would be fine for that?